bootstrap/kubeadm/internal/cloudinit/node_test.go

```diff
@@ -0,0 +1,87 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package cloudinit
+
+import (
+	"fmt"
+	"testing"
+
+	bootstrapv1 "sigs.k8s.io/cluster-api/bootstrap/kubeadm/api/v1alpha4"
+	"sigs.k8s.io/yaml"
+)
+
+func TestNewNode(t *testing.T) {
+	tests := []struct {
+		name    string
+		input   *NodeInput
+		check   func([]byte) error
+		wantErr bool
+	}{
+		{
+			"check for duplicated WriteFiles",
+			&NodeInput{
+				BaseUserData: BaseUserData{
+					AdditionalFiles: []bootstrapv1.File{
+						{
+							Path:        "/etc/foo.conf",
+							Content:     "bar",
+							Owner:       "root",
+							Permissions: "0644",
+						},
+					},
+				},
+			},
+			checkDuplicateWriteFiles,
+			false,
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			got, err := NewNode(tt.input)
+			if (err != nil) != tt.wantErr {
+				t.Errorf("NewNode() error = %v, wantErr %v", err, tt.wantErr)
+				return
+			}
+			if err := tt.check(got); err != nil {
+				t.Errorf("%v: got = %s", err, got)
+			}
+		})
+	}
+}
+
+func checkDuplicateWriteFiles(b []byte) error {
```

Comment:
I know this is a little bit of a stretch, but what about making this a little bit more generic by checking that a given list of files is included in cloud init/write_files - without duplicates -, so we can later reuse this check for init/join control plane cloud init unit tests?

something like `checkWriteFiles(files string...) func (b []byte) error { ... }`

Reply:
@fabriziopandini I assume you did mean to 
* check all elements in cloudinitData.WriteFiles for duplicates 
* and check for existence of the given files

Is this correct?
I think it never makes sense to have any duplicate right?

Reply:
Or according the function name `checkWriteFiles` I think we only should check the given files for duplicates?

Reply:
I adjusted the implementation. PTAL :slightly_smiling_face:

Reply:
a small nit, but yay!